sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,600 @@
+execute-contract(
+  contract-identifier,
+  transaction-name,
+  sender-principal,
+  transaction-arguments)
+
+  returns: true or false if the transaction executed successfully
+```
+
+## Invocation and Static Analysis
+
+When processing a client transaction, a `blockstack-core` node will do
+one of two things, depending on whether that transaction is a contract
+function invocation, or is attempting to publish a new smart contract.
+
+### Contract function invocation
+
+Any transaction which invokes a smart contract will be included in the
+blockchain, so long as it pays a Stacks fee greater than the minimum
+fee. This is true even for transactions which are _invalid_. This is
+because _validating_ an invalid transaction is not a free operation.
+
+1. Get the balance of the sender's account. If it's less than the tx fee,
+then `RETURN INVALID`.
+2. Otherwise, debit the user's account by the tx fee.
+3. Look up the contract by hash. If it does not exist, then `RETURN INVALID`.
+4. Look up the contract's `define-public` function and compare the
+   tx's arguments against it. If the tx does not call an existing
+   method, or supplies invalid arguments, then `RETURN INVALID`.
+5. Look up the cost to execute the given function, and if it is greater
+   than the paid tx fee, `RETURN INVALID`.
+6. Execute the public function code and commit the effects of running the code and `RETURN OK`
+
+### Publish contract
+
+A transaction which creates a new smart contract must pay a fee which
+funds the static analysis required to determine the cost of the new
+smart contract's public functions. To process such a transaction,
+`blockstack-core` will:
+
+1. Check the sender's account balance. If zero, then `RETURN INVALID`
+2. Check the tx fee against the user's balance. If it's higher, then `RETURN INVALID`
+3. Debit the tx fee from the user's balance.
+4. Check the syntax, calculating the fee of verifying each code
+   item. If the cost of checking the next item exceeds the tx fee, or
+   if the syntax is invalid, then `RETURN INVALID`.
+5. Build the AST, and assign a fee for adding each AST item. If the
+   cost of adding the next item to the tree exceeds the tx fee (or if
+   the AST gets too big), then `RETURN INVALID`.
+6. Walk the AST. Each step in the walk incurs a small fee. Do the
+   following while the tx fee is higher than the total cost incurred
+   by walking to the next node in the AST:
+   a. If the next node calls a contract method, then verify that
+      the contract exists and the method arguments match the contract's
+      `define-public` signature. If not, then `RETURN INVALID`.
+   b. Compute the runtime cost of each node in the AST, adding it
+      to the function's cost analysis.
+7. Find all `define-map` calls to find all tables that need to
+   exist. Each step in this incurs a small fee.
+8. Create all the tables if the cost of creating them is smaller than
+   the remaining tx fee. If not, then RETURN INVALID.
+9. `RETURN OK`
+
```

Comment:
As described, I think there's a DoS vulnerability here:  I can send the network maximally-sized contracts that have a syntax error early on, and the entire contract body will be included in the chain (effectively getting underpriced storage).  Same goes for transactions that call a public contract function in an invalid way.

I think there needs to be a check on the length of code submitted by a transaction (either code to define a contract or to invoke a method), and that the sender needs to be assessed a storage fee before any validation occurs.  Transactions that don't include the storage fee are dropped from the mempool and not included at all.